community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/PlannerCache.scala

```diff
@@ -19,75 +19,61 @@
-}
+class PlannerFactory(graph: GraphDatabaseService, kernelAPI: KernelAPI, kernelMonitors: KernelMonitors, log: Log,
+                     queryCacheSize: Int, statisticsDivergenceThreshold: Double, queryPlanTTL: Long,
+                     useErrorsOverWarnings: Boolean) {
 
-trait PlannerCache[K <: PlannerSpec]  {
-  def apply[S](spec: K { type SPI = S }): S
-}
+  def create(spec: PlannerSpec_v1_9.type) = CompatibilityFor1_9(graph, queryCacheSize, kernelMonitors)
 
-object VersionBasedPlannerCache {
-
-  class PlannerMap[K <: PlannerSpec { type SPI = S }, S](factory: PlannerFactory) {
-    private val cell = newCell(newRef(Map.empty[K, S]))
-
-    def apply(spec: K): S = recurse(spec, None)
-
-    @tailrec
-    private final def recurse(spec: K, candidate: Option[S]): S = {
-      val ref = cell.get()
-      val map = ref.get.getOrElse(Map.empty)
-      map.get(spec) match {
-        case Some(result) => result
-        case None =>
-          val nextCandidate = candidate.getOrElse(factory.create(spec))
-          val nextRef = newRef(map.updated(spec, nextCandidate))
-          if (cell.compareAndSet(ref, nextRef)) nextCandidate else recurse(spec, candidate orElse Some(nextCandidate))
-      }
-    }
+  def create(spec: PlannerSpec_v2_2) = spec.planner match {
+    case CypherPlanner.rule => CompatibilityFor2_2Rule(graph, queryCacheSize, statisticsDivergenceThreshold,
+      queryPlanTTL, CypherCompiler.CLOCK, kernelMonitors, kernelAPI)
+    case _ => CompatibilityFor2_2Cost(graph, queryCacheSize, statisticsDivergenceThreshold, queryPlanTTL,
+      CypherCompiler.CLOCK, kernelMonitors, kernelAPI, log, spec.planner)
   }
 
-  private def newCell[T <: AnyRef](v: T) = new AtomicReference(v)
-  private def newRef[T <: AnyRef](v: T) = new WeakReference[T](v)
+  def create(spec: PlannerSpec_v2_3) =  spec.planner match {
+    case CypherPlanner.rule => CompatibilityFor2_3Rule(graph, queryCacheSize, statisticsDivergenceThreshold,
+      queryPlanTTL, CypherCompiler.CLOCK, kernelMonitors, kernelAPI)
+    case _ => CompatibilityFor2_3Cost(graph, queryCacheSize, statisticsDivergenceThreshold, queryPlanTTL,
+      CypherCompiler.CLOCK, kernelMonitors, kernelAPI, log, spec.planner, spec.runtime, useErrorsOverWarnings)
+  }
 }
 
-class VersionBasedPlannerCache(factory: PlannerFactory) extends PlannerCache[PlannerSpec] {
+class PlannerCache(factory: PlannerFactory)  {
+  private val cache_v1_9 = new CachingValue[CompatibilityFor1_9]
+  private val cache_v2_2 = new mutable.HashMap[PlannerSpec_v2_2, CompatibilityFor2_2]
+  private val cache_v2_3 = new mutable.HashMap[PlannerSpec_v2_3, CompatibilityFor2_3]
 
-  import VersionBasedPlannerCache.PlannerMap
+  def apply(spec: PlannerSpec_v1_9.type) = cache_v1_9.getOrElseUpdate(factory.create(spec))
+  def apply(spec: PlannerSpec_v2_2) = cache_v2_2.getOrElseUpdate(spec, factory.create(spec))
+  def apply(spec: PlannerSpec_v2_3) = cache_v2_3.getOrElseUpdate(spec, factory.create(spec))
```

Comment:
Couldn't these three be replaced by one function `apply(spec: PlannerSpec) = spec match { ... }` with pattern matching cases for the various subtypes, instead of overloading?